Assets/MixedRealityToolkit.SDK/Experimental/Features/Dwell/DwellHandler.cs

```diff
@@ -0,0 +1,243 @@
+                        return Mathf.Clamp((float)(dwellProfile.TimeToCompleteDwell - FillTimer) / dwellProfile.TimeToCompleteDwell, 0f, 1f);
+                    case DwellStateType.DwellCompleted:
+                        return 1;
+                    case DwellStateType.DwellCanceled:
+                        if ((DateTime.UtcNow - focusExitTime).TotalSeconds > dwellProfile.TimeToAllowDwellResume)
+                        {
+                            CurrentDwellState = DwellStateType.None;
+                            return 0;
+                        }
+                        else if (dwellProfile.TimeToAllowDwellResume > 0)
+                        {
+                            return Mathf.Clamp((float)(dwellProfile.TimeToCompleteDwell - FillTimer) / dwellProfile.TimeToCompleteDwell, 0f, 1f);
+                        }
+                        break;
+                    default:
+                        return 0;
+                }
+
+                return 0;
+            }
+        }
+
+        /// <summary>
+        /// Cached pointer reference to track focus events maps to teh same pointer that initiated dwell
+        /// </summary>
+        private IMixedRealityPointer pointer;
+
+        private DateTime focusEnterTime = DateTime.MaxValue;
+        private DateTime focusExitTime = DateTime.MaxValue;
+
+        /// <summary>
+        /// Exposes whether the target has focus from the pointer defined in dwell profile settings
+        /// </summary>
+        protected bool HasFocus { get; private set; }
+
+        /// <summary>
+        /// Captures the dwell status 
+        /// </summary>
+        protected DwellStateType CurrentDwellState = DwellStateType.None;
+
+        /// <summary>
+        /// Abstracted value for the how long the dwelled object still needs to be focused to comlete the dwell action
+        /// </summary>
+        protected float FillTimer = 0;
+
+        private void Awake()
+        {
+            Debug.Assert(dwellProfile != null, "DwellProfile is null, creating default profile.");
+
+            if (dwellProfile == null)
+            {
+                dwellProfile = ScriptableObject.CreateInstance<DwellProfile>();
+            }
+
+            CancelDwellAction += OnDwellCancelAction;
+        }
+
+        private void Update()
+        {
+            UpdateFillTimer();
+
+            if (HasFocus && CurrentDwellState != DwellStateType.DwellCompleted)
+            {
+                double focusDuration = (DateTime.UtcNow - this.focusEnterTime).TotalSeconds;
+
+                if (CurrentDwellState == DwellStateType.FocusGained && focusDuration >= dwellProfile.DwellIntentDelay)
+                {
+                    DwellIntended.Invoke(pointer);
+                    CurrentDwellState = DwellStateType.DwellIntended;
+                }
+                else if (CurrentDwellState == DwellStateType.DwellIntended && (focusDuration - dwellProfile.DwellIntentDelay) >= dwellProfile.DwellStartDelay)
```

Comment:
>(focusDuration - dwellProfile.DwellIntentDelay) [](start = 78, length = 47)

Nit: This section is interesting because it tracks the current time in the current phase - you don't need to do this, but from a readability perspective it would would easier to read to have something like:

GetTimeInCurrentState() - which would then have a cleaner/more readable thing in these if statements, and then also have a single helper where it's clear that each of these operations (i.e. focusDuration - whatever - whatever) is clear that it's how you calculate the time in each state.

Reply:
I feel this is clear enough... marking as Won't Fix unless you have strong feelings about this.